conceptnet_sme/semantic_matching.py

```diff
@@ -519,18 +537,322 @@ def export(self, dirname):
+            self.add_module('child:{}'.format(dev.index), module_copy)
+        module.to(device)
+        self.chunk_sizes = torch.zeros(len(self.children), dtype=torch.int64)
+
+    def forward(self, *args):
+        """
+        Scatter the supplied args (assumed to be a list of tensors) across 
+        the child modules, and gather their outputs (assumed to be single 
+        tensors) back to the first gpu.  Also, accumulate the sizes of the 
+        scattered chunks (for later use in updating parameter gradients).
+        """
+        if len(self.children) <= 1:
+            return self.children[0](*args)
+        device_ids = [device.index for device in self.devices]
+        chunk_lists = list(list() for i_device in device_ids)
+        for arg in args:
+            chunks = torch.cuda.comm.scatter(arg, device_ids)
+            for i_child, chunk in enumerate(chunks):
+                chunk_lists[i_child].append(chunk)
+        chunks = list(tuple(chunk_list) for chunk_list in chunk_lists)
+        outputs = []
+        for i_child, (module, chunk) in enumerate(zip(self.children, chunks)):
+            chunk_size = chunk[0].size()[0]
+            self.chunk_sizes[i_child] += chunk_size
+            output = module(*chunk)
+            outputs.append(output)
+        assert len(self.children) == len(outputs)
+        output = torch.cuda.comm.gather(
+            outputs, destination=self.devices[0].index)
+        return output
+
+    def zero_grad(self):
+        """
+        In addition to the ordinary zeroing of gradient data, reset the 
+        chunk size data.
+        """
+        super().zero_grad()
+        self.chunk_sizes = torch.zeros(len(self.children), dtype=torch.int64)
+
+    def broadcast_gradients(self):
+        """
+        Compute a single value, for all the child modules, of the gradient 
+        of each module parameter (as a convex combination of the gradients 
+        in the individual children, with coefficients based on the batch 
+        sizes from the forward computation), and distribute these common 
+        gradients back to all the children.
+        """
+        if len(self.children) <= 1:
+            return
+        weights = self.chunk_sizes.to(torch.float32) / \
+                  self.chunk_sizes.sum().to(torch.float32)
+        weights = [weights[ii].item() for ii, dev in
+                   enumerate(self.devices)]
+        for name, param in self.children[0].named_parameters():
+            if param.grad is None:
+                continue
+            param_copies = [param]
+            for other_module in self.children[1:]:
+                other_module_params = list(
+                    p for other_name, p in other_module.named_parameters()
+                    if other_name == name)
+                assert len(other_module_params) == 1
+                param_copies.append(other_module_params[0])
+            param_grad = torch.cuda.comm.reduce_add(
```

Comment:
This thing with `reduce_add` and `mul_` requires some explanation.

Reply:
I've added additional comments; if you find them insufficient please let me know.